kythe/cxx/indexer/cxx/IndexerASTHooks.cc

```diff
@@ -3827,856 +3992,606 @@ void IndexerASTVisitor::DumpTypeContext(unsigned Depth, unsigned Index) {
+
+NodeSet IndexerASTVisitor::BuildNodeSetForObjCTypeParam(
+    clang::ObjCTypeParamTypeLoc TL) {
+  if (const auto* Decl = TL.getDecl()) {
+    return BuildNodeIdForDecl(Decl);
+  }
+  return NodeSet::Empty();
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForAttributed(
+    clang::AttributedTypeLoc TL) {
+  return BuildNodeSetForType(TL.getModifiedLoc());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForDependentAddressSpace(
+    clang::DependentAddressSpaceTypeLoc TL) {
+  return BuildNodeSetForType(TL.getPointeeTypeLoc());
+}
+
+GraphObserver::NodeId IndexerASTVisitor::BuildNominalNodeIdForDecl(
+    const clang::NamedDecl* Decl) {
+  auto NominalID = Observer.nodeIdForNominalTypeNode(BuildNameIdForDecl(Decl));
+  return Observer.recordNominalTypeNode(
+      NominalID, MarkedSources.Generate(Decl).GenerateMarkedSource(NominalID),
+      GetDeclChildOf(Decl));
+}
+
+const clang::TemplateTypeParmDecl*
+IndexerASTVisitor::FindTemplateTypeParmTypeLocDecl(
+    clang::TemplateTypeParmTypeLoc TL) const {
+  // Either the `TemplateTypeParm` will link directly to a relevant
+  // `TemplateTypeParmDecl` or (particularly in the case of canonicalized
+  // types) we will find the Decl in the `Job->TypeContext` according to the
+  // parameter's depth and index.
+  // Depths count from the outside-in; each Template*ParmDecl has only
+  // one possible (depth, index).
+  if (auto* Decl = TL.getDecl()) {
+    return Decl;
+  }
+  LOG(INFO) << "Immediate TemplateTypeParmDecl not found, falling back to "
+               "TypeContext";
+  if (auto* TypeParm = TL.getTypePtr()) {
+    if (TypeParm->getDepth() < Job->TypeContext.size() &&
+        TypeParm->getIndex() < Job->TypeContext[TypeParm->getDepth()]->size()) {
+      return cast<clang::TemplateTypeParmDecl>(
+          Job->TypeContext[TypeParm->getDepth()]->getParam(
+              TypeParm->getIndex()));
+    }
+  }
+  LOG(ERROR)
+      << "Unable to find TemplateTypeParmDecl for TemplateTypeParmTypeLoc";
+  return nullptr;
+}
+
+absl::optional<NodeId> IndexerASTVisitor::BuildNodeIdForType(
+    const clang::TypeLoc& TypeLoc) {
+  return BuildNodeSetForType(TypeLoc).AsOptional();
+}
+
+absl::optional<NodeId> IndexerASTVisitor::BuildNodeIdForType(
+    const clang::QualType& QT) {
+  return BuildNodeSetForType(QT).AsOptional();
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForType(const clang::QualType& QT) {
+  // TODO(shahms): This be barkwards; we should build NodeSets from
+  // `clang::Type*` subclasses, rather than TypeLoc's with occasionally empty
+  // locations.
+  CHECK(!QT.isNull());
+  TypeSourceInfo* TSI = Context.getTrivialTypeSourceInfo(QT, SourceLocation());
+  return BuildNodeSetForType(TSI->getTypeLoc());
+}
+
+NodeSet IndexerASTVisitor::BuildNodeSetForType(const clang::TypeLoc& TL) {
+  TypeKey Key(Context, TL.getType(), TL.getTypePtr());
+  const auto& Prev = TypeNodes.find(Key);
+  if (Prev != TypeNodes.end()) {
+    return Prev->second;
+  }
+  NodeSet Nodes = [&]() -> NodeSet {
```

Comment:
Why is this an anonymous function?

Reply:
Because it keeps the logic inside the switch (and therefor the macros) simple and clean.  We also only initialize the NodeSet once, rather than default initialize-then-assign..